src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelEMAlgorithm.java

```diff
@@ -0,0 +1,497 @@
+package org.broadinstitute.hellbender.tools.coveragemodel;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.math3.util.FastMath;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.AlleleMetadataProducer;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.CallStringProducer;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.ScalarProducer;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * This class implements the EM algorithm for the GATK Bayesian coverage model. It is used for:
+ *
+ * (1) performing max likelihood estimation of the model parameters AND calculating posterior expectations
+ *     via calling {@link CoverageModelEMAlgorithm {@link #runExpectationMaximization()}}
+ *
+ * (2) Calculating posterior expectations for the model parameters already existing in the workspace
+ *     via calling {@link CoverageModelEMAlgorithm {@link #runExpectation()}}
+
+ * (see CNV-methods.pdf for technical details).
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+
+public final class CoverageModelEMAlgorithm<S extends AlleleMetadataProducer & CallStringProducer & ScalarProducer> {
```

Comment:
Sure. We need to sit down and think hard about refactoring... and it all depends on whether or not we need to elevate this to a general purpose framework.